userdocs/diagnostics/expression-too-complex.md

```diff
@@ -0,0 +1,16 @@
+# Type Inference in Complex Expressions
+For some complex expressions, the Swift compiler is unable to infer types in reasonable time. This can occur for a number of reasons, but some common ones include:
+
+- The expression makes use of many overloaded operators and/or functions.
+- The expression contains literals which can be used as values of many different types (e.g. integer and floating point literals).
+- The expression includes a collection literal with a large number of nontrivial element expressions.
+
+These types of expressions can take a long time to typecheck because the compiler must consider and eliminate a larger number of possible solutions when assigning types to subexpressions.
+
+When this error occurs, possible solutions include:
+
+- Breaking the expression up into multiple subexpressions by storing intermediate values as `let` constants. These simpler subexpressions can have their types inferred more quickly.
+- Providing manual type annotations for parts of the expression. This reduces the number of types the compiler must infer, and as a result the number of possible solutions it must consider.
```

Comment:
```suggestion
- Providing manual type annotations for parts of the expression. This reduces the number of possible solutions that the compiler must consider.
```

It is self-evidently true that annotating with an explicit type means that the compiler no longer has to infer an implicit type.

Users are likely to be familiar with using a colon to indicate an explicit type on the LHS of an assignment, but I think in general many are not so familiar with using the `as` operator to indicate an explicit type for _parts_ of an expression, so it could be helpful to provide a simple example:

> For example, you can write `let array = [1 as Int, 2, 3]`.

Reply:
In the example I added, I showed ways to fix the problem with either `as` or an initializer call, which I think is sufficient